crates/iceberg/src/expr/visitors/bound_predicate_visitor.rs

```diff
@@ -0,0 +1,317 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::expr::{BoundPredicate, BoundReference, PredicateOperator};
+use crate::spec::Datum;
+use crate::Result;
+use fnv::FnvHashSet;
+
+pub(crate) enum OpLiteral<'a> {
+    Single(&'a Datum),
+    Set(&'a FnvHashSet<Datum>),
+}
+
+/// A visitor for [`BoundPredicate`]s. Visits in post-order.
+pub trait BoundPredicateVisitor {
+    /// The return type of this visitor
+    type T;
+
+    /// Called after an `AlwaysTrue` predicate is visited
+    fn always_true(&mut self) -> Result<Self::T>;
+
+    /// Called after an `AlwaysFalse` predicate is visited
+    fn always_false(&mut self) -> Result<Self::T>;
+
+    /// Called after an `And` predicate is visited
+    fn and(&mut self, lhs: Self::T, rhs: Self::T) -> Result<Self::T>;
+
+    /// Called after an `Or` predicate is visited
+    fn or(&mut self, lhs: Self::T, rhs: Self::T) -> Result<Self::T>;
+
+    /// Called after a `Not` predicate is visited
+    fn not(&mut self, inner: Self::T) -> Result<Self::T>;
+
+    /// Called after visiting a UnaryPredicate, BinaryPredicate,
+    /// or SetPredicate. Passes the predicate's operator in all cases,
+    /// as well as the term and literals in the case of binary and aet
+    /// predicates.
+    fn op(
```

Comment:
Hi, @sdd How about we try rewrite it like [this](https://github.com/liurenjie1024/iceberg-rust/blob/renjie/bound_visitor/crates/iceberg/src/expr/visitors/bound_predicate_visitor.rs)

Reply:
Hi Renjie! Thanks for the comment. I'm not totally convinced with your suggestion - we're trying to get the best performance, but by passing Set literals through as an iterator, rather than passing a reference to the HashSet, it requires either collecting the iterator, resulting in an allocation, or iterating over potentially the whole iterator.

Reply:
I think generating an iterator for hashset will not introduce allocation, but you are right that user may need to iterating the whole iterator to implement their logic. Method in this pr seems kind of overdesign and counterintutive for user to implement. In this case I think your original design seems more intuitive to me, as @Fokko said adding operator maybe not too frequent. Sorry for the back and forth, I didn't expect this approach to be so complicated and introduce so much overhead. 

cc @Fokko @marvinlanhenke @Xuanwo What do you think?

Reply:
Ok - here's another PR with the original design. I'll let you guys choose which one to merge :-) https://github.com/apache/iceberg-rust/pull/334. 

Here's how the inclusive projection would end up looking with this design too: https://github.com/apache/iceberg-rust/pull/335

Reply:
My preference is the original design, which is  (confusingly) in the "alternative" PRs 😅

Reply:
I'm not sold on the iterator idea either, it looks very fancy, but iterating over a set feels awkward to me. 

> If you're not happy with this approach @Fokko, I'll wait until the three of you are in agreement and fix up to whatever you all decide on. I'm happy with the approach as-is after switching my original design to the approach proposed by @liurenjie1024 and @marvinlanhenke.

Well, I'm not the one to decide here, just weighing in my opinion :D I have a strong preference for the original approach as suggested in https://github.com/apache/iceberg-rust/pull/334 I think this leverages the visitor as much as possible, and also makes sure that all the operations are covered.

Reply:
I'm also +1 for the original approach after comparing these two approaches.